src/hotspot/share/jfr/support/jfrSymbolTable.cpp

```diff
@@ -236,59 +235,48 @@ traceid JfrSymbolTable::mark(uintptr_t hash, const char* str, bool leakp) {
 }
 
 /*
-* hidden classes symbol is the external name +
-* the address of its InstanceKlass slash appended:
-*   java.lang.invoke.LambdaForm$BMH/22626602
-*
-* caller needs ResourceMark
-*/
-
-uintptr_t JfrSymbolTable::hidden_klass_name_hash(const InstanceKlass* ik) {
-  assert(ik != nullptr, "invariant");
-  assert(ik->is_hidden(), "invariant");
-  const oop mirror = ik->java_mirror_no_keepalive();
-  assert(mirror != nullptr, "invariant");
-  return (uintptr_t)mirror->identity_hash();
-}
-
-static const char* create_hidden_klass_symbol(const InstanceKlass* ik, uintptr_t hash) {
-  assert(ik != nullptr, "invariant");
-  assert(ik->is_hidden(), "invariant");
+ * The hidden class symbol is the external name with the
+ * address of its Klass slash appended.
+ *
+ * This is done by replacing the '+' delimiter with a '/'
+ * in a copy of the already mangled hidden class name.
+ *
+ * "java.lang.invoke.LambdaForm$DMH+0x0000000037144c00"
+ *  becomes
+ * "java.lang.invoke.LambdaForm$DMH/0x0000000037144c00"
+ *
+ * Caller needs ResourceMark.
+ */
+static const char* create_hidden_klass_name(const Klass* k, uintptr_t hash) {
+  assert(k != nullptr, "invariant");
+  assert(k->is_hidden(), "invariant");
   assert(hash != 0, "invariant");
-  char* hidden_symbol = nullptr;
-  const oop mirror = ik->java_mirror_no_keepalive();
-  assert(mirror != nullptr, "invariant");
-  char hash_buf[40];
-  os::snprintf_checked(hash_buf, sizeof(hash_buf), "/" UINTX_FORMAT, hash);
-  const size_t hash_len = strlen(hash_buf);
-  const size_t result_len = ik->name()->utf8_length();
-  hidden_symbol = NEW_RESOURCE_ARRAY(char, result_len + hash_len + 1);
-  ik->name()->as_klass_external_name(hidden_symbol, (int)result_len + 1);
-  assert(strlen(hidden_symbol) == result_len, "invariant");
-  strcpy(hidden_symbol + result_len, hash_buf);
-  assert(strlen(hidden_symbol) == result_len + hash_len, "invariant");
-  return hidden_symbol;
-}
-
-bool JfrSymbolTable::is_hidden_klass(const Klass* k) {
+  const Symbol* const name = k->name();
+  assert(name != nullptr, "invariant");
+  assert(name->identity_hash() == hash, "invariant");
+  const size_t len = static_cast<size_t>(name->utf8_length());
+  char* hidden_klass_name = NEW_RESOURCE_ARRAY(char, len + 1);
+  strncpy(hidden_klass_name, name->as_klass_external_name(), len + 1);
```

Comment:
Can't you use the `k->external_name()` which performs this plus to slash replacement already?

Reply:
Good point, turns out this support already exists.